File: bench/revcomp/revcomp.scala
/* The Computer Language Benchmarks Game
   http://shootout.alioth.debian.org/
   contributed by Eric Willigers
*/

object revcomp extends java.io.ByteArrayOutputStream with Application {
   val cmp = Array.range(0, 128)
   for ((a, b) <- "TAGCVHRMYKBDU".toList.zip("ATCGBDYKRMVHA".toList)) {
      cmp(a) = b
      cmp(a + 32) = b
   }

   def reverse() {
      if (count > 0) {
         var begin = buf.indexOf('\n') + 1
         var end = count - 1
         while (begin <= end) {
            if (buf(begin) == '\n') begin += 1
            if (buf(end) == '\n') end -= 1
            if (begin <= end) {
               val tmp = buf(begin)
               buf(begin) = cmp(buf(end)).byteValue
               begin += 1
               buf(end) = cmp(tmp).byteValue
               end -= 1
            }
         }
         System.out.write(buf, 0, count)
      }
   }

   val line = new Array[Byte](82)
   var read = System.in.read(line)
   while (read != -1) {
      var i = 0
      var last = 0
      while (i < read) {
         if (line(i) == '>') {
            write(line, last, i - last)
            reverse()
            reset()
            last = i
         }
         i += 1
      }
      write(line, last, read - last)
      read = System.in.read(line)
   }
   reverse()
}

File: bench/revcomp/revcomp.go
/* The Computer Language Benchmarks Game
 * http://shootout.alioth.debian.org/
 *
 * contributed by The Go Authors.
 */

package main

import (
   "bufio"
   "os"
)

const lineSize = 60

var complement = [256]uint8{
   'A': 'T', 'a': 'T',
   'C': 'G', 'c': 'G',
   'G': 'C', 'g': 'C',
   'T': 'A', 't': 'A',
   'U': 'A', 'u': 'A',
   'M': 'K', 'm': 'K',
   'R': 'Y', 'r': 'Y',
   'W': 'W', 'w': 'W',
   'S': 'S', 's': 'S',
   'Y': 'R', 'y': 'R',
   'K': 'M', 'k': 'M',
   'V': 'B', 'v': 'B',
   'H': 'D', 'h': 'D',
   'D': 'H', 'd': 'H',
   'B': 'V', 'b': 'V',
   'N': 'N', 'n': 'N',
}

func main() {
   in := bufio.NewReader(os.Stdin)
   buf := make([]byte, 1024*1024)
   line, err := in.ReadSlice('\n')
   for err == nil {
      os.Stdout.Write(line)

      // Accumulate reversed complement in buf[w:]
      nchar := 0
      w := len(buf)
      for {
         line, err = in.ReadSlice('\n')
         if err != nil || line[0] == '>' {
            break
         }
         line = line[0 : len(line)-1]
         nchar += len(line)
         if len(line)+nchar/60+128 >= w {
            nbuf := make([]byte, len(buf)*5)
            copy(nbuf[len(nbuf)-len(buf):], buf)
            w += len(nbuf) - len(buf)
            buf = nbuf
         }

         // This loop is the bottleneck.
         for _, c := range line {
            w--
            buf[w] = complement[c]
         }
      }

      // Copy down to beginning of buffer, inserting newlines.
      // The loop left room for the newlines and 128 bytes of padding.
      i := 0
      for j := w; j < len(buf); j += 60 {
         n := copy(buf[i:i+60], buf[j:])
         buf[i+n] = '\n'
         i += n + 1
      }
      os.Stdout.Write(buf[0:i])
   }
}
